### ISSUE TITLE
[Feature Request] Set cosmetics via file akin to plandomizer

### ISSUE THREAD

**Issue by dewiniaid** [TestRunnerSRL/OoT-Randomizer#1033]
I'm toying around with the concept of making a randomizer ROM using custom music assets.  I managed to get that working (using the existing instructions about files in `data/Music`) but quickly learned that there's not a way to explicitly assign a specific song to a specific area.  With the advent of Plando, a handcrafted run can be even more entertaining with some tailored changes.  (For instance, using cleverly crafted remixes of Bowser's Castle for Ganon's Castle.)

I am thinking the end result could look something like this:
- Output a json version of chosen cosmetics/SFX in the spoiler log in addition to the human-readable cosmetics log.  (As a bonus, this makes a list of randomized cosmetics available to people using the online randomizer.)
- Allow the existing system for the plandomizer to specify cosmetics as well.
- For a purely random run using a "cosmetics" plan, allow a second opportunity to specify a .json file that will _only_ affect cosmetics.  This is mostly for the use case of runners wanting a truly random run (including one from a seed already created by someone else) but wanting to explicitly assign music/SFX).  The file format for this would be the same as that used by Plandomizer, with all non-cosmetic options ignored.

I've also learned that when generating .zpf files, the randomizer doesn't support custom music assets and will only choose from those available in-game -- this significantly complicates distributing a seed with custom music as the only "legal" option for doing so is creating and then sending a BPS patch.  Including the music in the cosmetics json could solve this, with something ala:

```json
"music": {
	"Market": "Zoras Domain",
	"Title Theme": {
		"name": "Something Fancy and Special",
		"audiobank": 6,
		"data": "base64-encoded Something Fancy and Special.seq goes here"
	}
}
```

(`audiobank` is pulled from the .meta file in this case.)

**Comment by flagrama** [comment]
While I have liked the idea of this since it was first brought up, including the actual copyrighted data in any way that is to be distributed by the randomizer will not happen. A BPS patch with custom music would also not be legal AFAIK, as it needs to include the whole of the copyrighted asset.

**Comment by dewiniaid** [comment]
I didn't consider the possibility of Plandos being saved, good point.

Using the logic here of "Assume the user has permission to distribute, but don't host anything so we're not liable if they don't".  Or, taking it further, let's say I made a Plando using some public domain music I remixed and use some hint stones for attribution:

Custom music probably shouldn't be allowed in a plando file that's available on https://ootrandomizer.com/, true.  Users lie about "I made this and this is totally legitimate original content" all the time.

As I understand it (I could be wrong!), ZPF files aren't saved at all.  Allowing custom music to be embedded in a ZPF file then removes https://ootrandomizer.com/ from the distribution equation altogether -- just like it requires that you obtain an OOT ROM somewhere, custom music would require you'd obtain the ZPF with that music data already embedded somewhere.

**Comment by flagrama** [comment]
We cannot assume the user has permission. They almost certainly don't. We already are walking on a thin line with software that modifies existing code. We keep any possibility that we will be confused for distributing copyrighted content away entirely. There is no room for the risk that assumption would bring.

ZPF files are saved on the server. They are used for patching the ROMs. Nothing in a ZPF patch is copyrighted content by anyone but project members.

Custom music already requires you obtain it somewhere else. I don't see a problem with having a user using a cosmetics plando that has some known custom music known by name in it, that the offline randomizer then checks for in the proper location, but embedding them is a no go IMO. People will share them on the discord or other randomizer-related communities, then we will have to ban them for distributing copyrighted content.

We also prefer to keep as many people using the website as possible. It's much easier to support our single known configuration compared to thousands of potential different ones. We usually don't tell people to use the offline version unless they are creating a complicated plando that may take several generations to succeed (this one is just a personal suggestion, the website owners don't mind them doing this on the website), they want to use custom music, or the website is down.

Ideally, custom music will be a little bit easier to deal with in the future, probably after we figure out a system for custom models, since that is unlikely to be as tied to the randomizer code as the initial implementation of custom music. The user will provide any content they wish to use when patching so that nothing touches the patch file, only the finalized ROM that gets generated only locally.

**Comment by darkangel93x** [comment]
Well i would love to see a plando option happen. (i have a bunch of ideas how to use that properly)
For vanilla music that should be no big issue even on the online version.

A plando with custom music would, if it happens, ONLY working offline though.

**Comment by flagrama** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Cosmetics Plandomizer](https://trello.com/c/BIea5MQ9/20-cosmetics-plandomizer)

**Comment by ETR-BTF** [comment]
A cosmetic plandomizer feature was implemented in v6.0 of the randomizer, released in December 2020. :)
